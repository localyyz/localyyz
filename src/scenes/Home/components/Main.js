import React from "react";
import { View, StyleSheet, FlatList, StatusBar } from "react-native";

// custom
import { NAVBAR_HEIGHT, Colours, Sizes } from "localyyz/constants";
import { ReactiveSpacer } from "localyyz/components";

// third party
import { withNavigation } from "react-navigation";
import { observer, inject } from "mobx-react/native";
import LinearGradient from "react-native-linear-gradient";

// local
import { Banner, Collection, Brands } from "../blocks";
import MainPlaceholder from "./MainPlaceholder";

// constants
const VIEWABLITY_CONFIG = {
  viewAreaCoveragePercentThreshold: 40
};

@inject(stores => ({
  homeStore: stores.homeStore,
  onScrollAnimate: stores.homeStore.onScrollAnimate,

  // blocks
  blocks: stores.homeStore.blocks,
  onViewableBlockChange: changes =>
    stores.homeStore.onViewableBlockChange(changes),
  updateBlockHeight: (i, evt) => stores.homeStore.updateBlockHeight(i, evt),

  // block fetching
  fetchCollectionBlocks: () => stores.homeStore.fetchCollectionBlocks()
}))
@observer
export class Main extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLayoutReady: false
    };

    // bindings
    this.scrollTo = this.scrollTo.bind(this);
    this.renderBlock = this.renderBlock.bind(this);
    this.onViewableBlockChange = this.onViewableBlockChange.bind(this);
    this.load = this.load.bind(this);
  }

  componentDidMount() {
    this.load();
  }

  async load() {
    //await this.props.fetchCategoryBlocks();
    await this.props.fetchCollectionBlocks();

    // and finally allow render
    this.setState({ isLayoutReady: true });
  }

  onViewableBlockChange(changes) {
    return this.props.onViewableBlockChange(changes);
  }

  scrollTo(n) {
    this.refs.blocks
      && this.refs.blocks.scrollToIndex({
        index: n,
        viewOffset: Sizes.Height / 8,
        viewPosition: 0
      });
  }

  get renderBlocks() {
    return (
      <FlatList
        ref="blocks"
        data={this.props.blocks}
        keyExtractor={block => `block-${block.id}`}
        contentContainerStyle={styles.content}
        renderItem={this.renderBlock}
        showsVerticalScrollIndicator={false}
        scrollEventThrottle={16}
        onViewableItemsChanged={this.onViewableBlockChange}
        viewabilityConfig={VIEWABLITY_CONFIG}
        onScroll={this.props.onScrollAnimate}/>
    );
  }

  renderBlock({ item: block, index: i }) {
    let component;
    switch (block.type) {
      case "header":
        component = (
          <ReactiveSpacer
            id={i}
            offset={Sizes.StatusBar}
            store={this.props.homeStore}
            heightProp="headerHeight"/>
        );
        break;
      case "productList":
        component = (
          <Collection
            {...block}
            withMargin
            id={i}
            fetchFrom={block.path}
            basePath={block.path}/>
        );
        break;
      case "collection":
        component = (
          <View
            style={i > 0 ? styles.blockContainer : styles.firstBlockContainer}>
            <Banner {...block} id={i} imageUri={block.imageUrl} />
            <Collection
              {...block}
              withMargin
              noMargin
              hideHeader
              fetchFrom={block.path}
              limit={4}/>
          </View>
        );
        break;
      case "brand":
        component = (
          <Brands
            {...block}
            id={i}
            limit={12}
            type={block.brandType}
            shouldShowName={!!block.shouldShowName}/>
        );
        break;
    }

    return (
      <View onLayout={evt => this.props.updateBlockHeight(i, evt)}>
        {component}
      </View>
    );
  }

  render() {
    return this.state.isLayoutReady ? (
      <View style={styles.container}>
        <View style={styles.statusBar} />
        <View style={styles.contentContainer}>{this.renderBlocks}</View>
        <View pointerEvents="box-none" style={styles.filter}>
          <LinearGradient
            colors={[Colours.WhiteTransparent, Colours.Transparent]}
            start={{ x: 0, y: 1 }}
            end={{ x: 0, y: 0 }}
            style={styles.gradient}
            pointerEvents="box-none"/>
        </View>
      </View>
    ) : (
      <MainPlaceholder />
    );
  }
}

export default withNavigation(Main);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colours.Foreground
  },

  statusBar: {
    height: Sizes.StatusBar,
    backgroundColor: Colours.Foreground
  },

  gradient: {
    alignItems: "center",
    justifyContent: "flex-end",
    height: Sizes.Height / 7,
    width: Sizes.Width
  },

  contentContainer: {
    flex: 1
  },

  filter: {
    position: "absolute",
    bottom: 0,
    left: 0,
    right: 0
  },

  content: {
    backgroundColor: Colours.Background,
    paddingBottom: NAVBAR_HEIGHT
  },

  blockContainer: {
    marginVertical: Sizes.InnerFrame / 2
  },

  firstBlockContainer: {
    marginBottom: Sizes.InnerFrame / 2
  }
});
