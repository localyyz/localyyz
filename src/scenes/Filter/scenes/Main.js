import React from "react";
import { View, StyleSheet, TouchableOpacity } from "react-native";

// custom
import { Styles, Colours, Sizes, NAVBAR_HEIGHT } from "localyyz/constants";
import { BaseScene, Filter } from "localyyz/components";

// temp (TODO: move this into its own component lib)
import { ProductCount } from "../../../components/Filter";

// third party
import { observer, Provider } from "mobx-react/native";
import LinearGradient from "react-native-linear-gradient";

@observer
export default class FilterMain extends React.Component {
  constructor(props) {
    super(props);

    // bindings
    this.close = this.close.bind(this);
  }

  get settings() {
    return (
      (this.props.navigation
        && this.props.navigation.state
        && this.props.navigation.state.params)
      || this.props
    );
  }

  componentDidMount() {
    this.settings.store && this.settings.store.refresh();
  }

  close() {
    return this.settings.onBack
      ? this.settings.onBack()
      : this.props.navigation.goBack(null);
  }

  render() {
    return (
      <Provider filterStore={this.settings.store}>
        <View style={styles.container}>
          <BaseScene
            iconType="close"
            backAction={this.close}
            title="Filter"
            style={styles.content}>
            <Filter style={styles.filter} />
          </BaseScene>
          <View pointerEvents="box-none" style={styles.footer}>
            <LinearGradient
              pointerEvents="box-none"
              colors={[Colours.WhiteTransparent, Colours.Transparent]}
              start={{ y: 1, x: 0 }}
              end={{ y: 0, x: 0 }}
              style={styles.gradient}>
              <View style={styles.toggle}>
                <TouchableOpacity onPress={this.close}>
                  <View style={styles.button}>
                    <ProductCount labelStyle={styles.label} />
                  </View>
                </TouchableOpacity>
              </View>
            </LinearGradient>
          </View>
        </View>
      </Provider>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginBottom: Sizes.ScreenBottom + NAVBAR_HEIGHT
  },

  content: {
    backgroundColor: Colours.Foreground,
    paddingTop: Sizes.InnerFrame
  },

  footer: {
    position: "absolute",
    bottom: 0,
    left: 0,
    right: 0
  },

  gradient: {
    paddingBottom: Sizes.ScreenBottom
  },

  toggle: {
    alignItems: "center"
  },

  button: {
    ...Styles.RoundedButton,
    alignItems: "center",
    margin: Sizes.InnerFrame,
    paddingHorizontal: Sizes.OuterFrame * 2,
    backgroundColor: Colours.Action
  },

  label: {
    ...Styles.Text,
    ...Styles.Emphasized
  },

  filter: {
    paddingHorizontal: Sizes.InnerFrame
  }
});
