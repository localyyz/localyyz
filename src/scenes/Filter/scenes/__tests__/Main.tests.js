import React from "react";

// custom
import ShallowRenderer from "react-test-renderer/shallow";

// local
import { default as Component } from "../Main";
let shallowRenderer = new ShallowRenderer();

// constants
const NAME = "Filter Main";
const PROPS = {
  store: {}
};

describe(NAME, () => {
  it("should render properly", () => {
    shallowRenderer.render(<Component {...PROPS} />);
    const c = shallowRenderer.getRenderOutput();

    expect(c).toBeDefined();
  });

  it("should match snapshot", () => {
    shallowRenderer.render(<Component {...PROPS} />);
    const c = shallowRenderer.getRenderOutput();

    expect(c).toMatchSnapshot();
  });
});
