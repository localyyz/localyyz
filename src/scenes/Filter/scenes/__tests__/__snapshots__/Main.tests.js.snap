// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Main should match snapshot 1`] = `
<Provider
  filterStore={Object {}}
>
  <Component
    style={
      Object {
        "flex": 1,
        "marginBottom": 48,
      }
    }
  >
    <BaseScene
      backAction={[Function]}
      iconType="close"
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "paddingTop": 15,
        }
      }
      title="Filter"
    >
      <Filter
        style={
          Object {
            "paddingHorizontal": 15,
          }
        }
      />
    </BaseScene>
    <Component
      pointerEvents="box-none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <LinearGradient
        colors={
          Array [
            "rgba(255, 255, 255, 0.9)",
            "rgba(255, 255, 255, 0)",
          ]
        }
        end={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        pointerEvents="box-none"
        start={
          Object {
            "x": 0,
            "y": 1,
          }
        }
        style={
          Object {
            "paddingBottom": 0,
          }
        }
      >
        <Component
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
          >
            <Component
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#DEEFF5",
                  "borderRadius": 15,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "margin": 15,
                  "paddingHorizontal": 50,
                  "paddingVertical": 7.5,
                }
              }
            >
              <inject-ProductCount
                labelStyle={
                  Object {
                    "backgroundColor": "rgba(255, 255, 255, 0)",
                    "color": "#000000",
                    "fontSize": 16,
                    "fontWeight": "600",
                  }
                }
              />
            </Component>
          </TouchableOpacity>
        </Component>
      </LinearGradient>
    </Component>
  </Component>
</Provider>
`;
